File: src/main/java/com/test/java/business/template/WordCountBolt.java
package com.test.java.business.template;

import backtype.storm.topology.BasicOutputCollector;
import backtype.storm.topology.OutputFieldsDeclarer;
import backtype.storm.topology.base.BaseBasicBolt;
import backtype.storm.tuple.Fields;
import backtype.storm.tuple.Tuple;
import backtype.storm.tuple.Values;

import java.util.HashMap;
import java.util.Map;

/**
 * @author lvxw
 */
public class WordCountBolt extends BaseBasicBolt {
    Map<String,Long> wordCountMap;
    Long initTime;

    public WordCountBolt() {
        wordCountMap = new HashMap<>();
        initTime = System.currentTimeMillis()/1000/5*5;
    }

    @Override
    public void execute(Tuple input, BasicOutputCollector collector) {
        String word = input.getStringByField("WORD");
        if(wordCountMap.get(word) == null){
            wordCountMap.put(word,1L);
        }else{
            wordCountMap.put(word,wordCountMap.get(word)+1);
        }

        Long currentTime = System.currentTimeMillis()/1000/5*5;
        if(currentTime > initTime){
            for(Map.Entry<String,Long> entry:wordCountMap.entrySet()){
                String value = entry.getKey()+"->"+entry.getValue();
                collector.emit(new Values(currentTime,value));
            }

            initTime = currentTime;
            wordCountMap.clear();
        }
    }

    @Override
    public void declareOutputFields(OutputFieldsDeclarer declarer) {
        declarer.declare(new Fields("TIME","VALUE"));
    }
}

File: src/main/scala/com/test/scala/business/template/WordCountBolt.scala
package com.test.scala.business.template


import backtype.storm.topology.{BasicOutputCollector, OutputFieldsDeclarer}
import backtype.storm.topology.base.BaseBasicBolt
import backtype.storm.tuple.{Fields, Tuple, Values}

import scala.collection.mutable

class WordCountBolt extends BaseBasicBolt{
  val wordCountMap = new mutable.HashMap[String, Long]
  var initTime = System.currentTimeMillis / 1000 / 5 * 5

  override def execute(input: Tuple, collector: BasicOutputCollector): Unit = {
    val word = input.getStringByField("WORD")
    wordCountMap += (word -> (wordCountMap.getOrElse(word,0L)+1L))

    val currentTime = System.currentTimeMillis / 1000 / 5 * 5
    if (currentTime > initTime) {
      for ((word,count) <- wordCountMap) {
        collector.emit(new Values(currentTime.toString, s"${word}->${count}"))
      }
      initTime = currentTime
      wordCountMap.clear()
    }
  }

  override def declareOutputFields(declarer: OutputFieldsDeclarer): Unit = {
    declarer.declare(new Fields("TIME", "VALUE"))
  }
}
